src/bin/ion/commands/generate/templates/java/class.templ

```diff
@@ -39,6 +39,45 @@ import java.io.IOException;
         }
     {% endfor %}
 
+    public static class {{ model.name }}Builder {
+        {% for field_name, field_val in struct_info["fields"] -%}
+            {% set field_value = field_val.0 | fully_qualified_type_name %}
+            private java.util.Optional<{{ field_value | wrapper_class }}> {{ field_name | camel }};
+        {% endfor %}
+
+        {{ model.name }}Builder() {
+            {# Initializes all the fields of this class #}
+            {% for field_name, field_val in struct_info["fields"] -%}
+                {% set field_value = field_val.0 | fully_qualified_type_name %}
+                {% set field_occurrence = field_val.1 %}
+                {# Initialize all the required fields as `java.util.Optional` to verify if the required fields were missing or not #}
+                this.{{ field_name | camel }} = java.util.Optional.empty();
+            {% endfor %}
+        }
+
+        {% for field_name, field_val in struct_info["fields"] %}
+            {% set field_value = field_val.0 %}
+            {% set val = field_value | fully_qualified_type_name %}
+            public {{ model.name }}Builder {{ field_name | camel }}({{ val }} {{ field_name | camel }}) {
+                this.{{ field_name | camel }} = java.util.Optional.of({{ field_name | camel }});
+                return this;
+            }
+        {% endfor %}
+
+        public {{ model.name }} build() {
+            {{ model.name }} {{ model.name | camel }} = new {{ model.name }}();
+            {% for field_name, field_val in struct_info["fields"] -%}
+                {# field_val.1 is the field occurrence #}
+                {% if field_val.1 == "Required" %}
+                if (!{{ field_name | camel }}.isPresent()) {
+                    throw new IonException("Can not find field name: {{ field_name }} for {{ model.name }} while reading Ion data.");
```

Comment:
The builder can also be used directly, by users. It should have a more generic exception so as not to mislead users about the source of the problem.
```suggestion
                    throw new IllegalArgumentException("No value provided for required field {{ field_name | camel }}");
```

Reply:
Good catch! Will change that.